src/search.rs

```diff
@@ -0,0 +1,79 @@
+use std::fmt::Write;
+
+use anyhow::Result;
+use itertools::Itertools;
+use serde::Deserialize;
+use textwrap_macros::dedent;
+
+#[derive(Deserialize)]
+pub struct SearchQuery {
+    username: Option<String>,
+    content: String,
+}
+
+impl SearchQuery {
+    pub fn build(self) -> Result<(String, Vec<String>)> {
+        let mut query = String::new();
+        let mut params = vec![];
+
+        write!(
+            query,
+            "{}",
+            dedent!(
+                r#"
+                SELECT messages.content, messages.username, messages.avatar, messages.sent_at
+                FROM messages_fts JOIN messages ON messages_fts.messages_rk = messages.rowid
+                WHERE
+                "#
+            )
+        )?;
+
+        if let Some(username) = self.username.map(fts_query) {
+            writeln!(query, r#"messages_fts.username MATCH ?{} OR"#, params.len() + 1)?;
+            params.push(format!("*\"{username}\"*"));
+        }
+
+        writeln!(query, r#"messages_fts.content MATCH ?{};"#, params.len() + 1)?;
+        params.push(fts_query(self.content));
+
+        Ok((query, params))
+    }
+}
+
+// Risks of injection are prevented by whitelisting only alphanumeric characters
```

Comment:
Can we not use prepared statements to safeguard against injection instead?
I feel like it could make sense to search for something that is not alphanumeric so a more general solution would be preferred.

Reply:
We are already using prepared statements in the actual query, but FTS5 `MATCH` clauses have their own [syntax](https://sqlite.org/fts5.html#full_text_query_syntax) and a malformed query like this can blow up:

```sql
SELECT * FROM messages_fts
WHERE content MATCH '*foo*';
```
For context, a more complex FTS query looks like this:

```sql
SELECT * FROM messages_fts
WHERE content MATCH '"franker"* AND "zambino" AND "something else"';
```

Literal strings in this syntax are tokenized by the [engine](https://sqlite.org/fts5.html#tokenizers) anyway, so I don't think there's harm in filtering things out. We may want to look into what tokenizers exactly do and see if we can provide a similar charset, but otherwise I don't think allowing non-alphanumeric would change the search results significantly.

Reply:
Fair enough, probably better ways we could structure the input instead then but for now it should be good 👍